src/ebmc/bmc.cpp

```diff
@@ -17,9 +17,181 @@ Author: Daniel Kroening, dkr@amazon.com
 #include <chrono>
 #include <fstream>
 
+void bmc_with_assumptions(
+  std::size_t bound,
+  const transition_systemt &transition_system,
+  ebmc_propertiest &properties,
+  decision_proceduret &solver,
+  message_handlert &message_handler)
+{
+  messaget message(message_handler);
+  const namespacet ns(transition_system.symbol_table);
+
+  // Use assumptions to check the properties separately
+
+  for(auto &property : properties.properties)
+  {
+    if(property.is_disabled() || property.is_failure() || property.is_assumed())
+    {
+      continue;
+    }
+
+    message.status() << "Checking " << property.name << messaget::eom;
+
+    auto assumption = not_exprt{conjunction(property.timeframe_handles)};
+
+    decision_proceduret::resultt dec_result = solver(assumption);
+
+    switch(dec_result)
+    {
+    case decision_proceduret::resultt::D_SATISFIABLE:
+      if(property.is_exists_path())
+      {
+        property.proved();
+        message.result() << "SAT: path found" << messaget::eom;
+      }
+      else // universal path property
+      {
+        property.refuted();
+        message.result() << "SAT: counterexample found" << messaget::eom;
+      }
+
+      property.witness_trace = compute_trans_trace(
+        property.timeframe_handles,
+        solver,
+        bound + 1,
+        ns,
+        transition_system.main_symbol->name);
+      break;
+
+    case decision_proceduret::resultt::D_UNSATISFIABLE:
+      if(property.is_exists_path())
+      {
+        message.result() << "UNSAT: No path found within bound"
+                         << messaget::eom;
+        property.refuted_with_bound(bound);
+      }
+      else // universal path property
+      {
+        message.result() << "UNSAT: No counterexample found within bound"
+                         << messaget::eom;
+        property.proved_with_bound(bound);
+      }
+      break;
+
+    case decision_proceduret::resultt::D_ERROR:
+      message.error() << "Error from decision procedure" << messaget::eom;
+      property.failure();
+      break;
+
+    default:
+      property.failure();
+      throw ebmc_errort() << "Unexpected result from decision procedure";
+    }
+  }
+}
+
+void bmc_without_assumptions(
+  std::size_t bound,
+  const transition_systemt &transition_system,
+  ebmc_propertiest &properties,
+  decision_proceduret &solver,
+  message_handlert &message_handler)
+{
+  messaget message(message_handler);
+  const namespacet ns(transition_system.symbol_table);
+
+  auto trace_found = [&solver](const ebmc_propertiest::propertyt &property)
+  {
+    for(auto &h : property.timeframe_handles)
+      if(solver.get(h).is_false())
+        return true;
+    return false;
```

Comment:
How about return solver.get(h).is_false();

Reply:
That won't work -- note the loop.